### PR TITLE
Implement gRPC streaming support for Java

### PR DESCRIPTION
**Pull request by shinfan**
- Implement surface changes corresponding to the GAX PR: https://github.com/googleapis/gax-java/pull/113
- Add Java code sample support for gRPC streaming
- Separate PageStreamingConfig with GrpcStreamingConfig
- Update baseline tests

### CODE REVIEW

**Comment by shinfan** [comment]
@jmuk I separate the grpc-streaming config from the page-streaming config since I feel using one class for both purposes can be confusing.

File: src/main/java/com/google/api/codegen/transformer/SurfaceTransformerContext.java [hunk 1]
```diff
@@ -117,6 +107,13 @@ private boolean isSupported(Method method) {
     return methods;
   }
 
+  private boolean isSupported(Method method) {
+    if (!getFeatureConfig().enableStreaming()) {
+      return !method.getRequestStreaming() && !method.getResponseStreaming();
+    }
+    return true;
+  }
```
Comment (jmuk) [inline src/main/java/com/google/api/codegen/transformer/SurfaceTransformerContext.java]: Not sure how worthwhile doing (and/or how does that fit with the design principle of MVVM), but can we just move `isSupported` into the FeatureConfig -- with receiving `Method` as an argument?

File: src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+  private final StreamingType type;
+
+  /**
+   * Creates an instance of GrpcStreamingConfig for gRPC response streaming, based on
+   * PageStreamingConfigProto, linking it up with the provided method. On errors, null will be
+   * returned, and diagnostics are reported to the diag collector.
+   */
+  @Nullable
+  public static GrpcStreamingConfig createGrpcStreaming(
+      DiagCollector diagCollector, PageStreamingConfigProto pageStreaming, Method method) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java]: Why does grpc streaming dake PageStreamingConfigProto as input?

File: src/main/java/com/google/api/codegen/config/MethodConfig.java [hunk 3]
```diff
@@ -334,10 +329,14 @@ public boolean hasRequestObjectMethod() {
...
-   * Returns the Interface that should be used for the GRPC call in place of the interface
-   * in which this method appears.
+   * Returns the Interface that should be used for the GRPC call in place of the interface in which
+   * this method appears.
    */
   public String getRerouteToGrpcInterface() {
     return rerouteToGrpcInterface;
   }
+
+  static private boolean isStreamingMethod(Method method) {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/MethodConfig.java]: why is this package-private?

File: src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java [hunk 4]
```diff
@@ -79,7 +80,11 @@ public ApiCallableTransformer() {
     apiCallableBuilder.memberName(context.getNamer().getSettingsMemberName(method));
     apiCallableBuilder.settingsFunctionName(context.getNamer().getSettingsFunctionName(method));
 
-    if (methodConfig.isBundling()) {
+    if (methodConfig.isGrpcStreaming()) {
+      apiCallableBuilder.type(ApiCallableType.StreamingApiCallable);
+      apiCallableBuilder.streamingType(methodConfig.getGrpcStreaming().getType());
+      typeTable.saveNicknameFor("io.grpc.stub.StreamObserver");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java]: This is Java-specific, so it should either be made more generic (go through SurfaceNamer) or should be moved somewhere else. (Likewise for the same import further down)

File: src/main/java/com/google/api/codegen/transformer/SurfaceTransformerContext.java (hunk 1 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/SurfaceTransformerContext.java]: By looking up the history, I followed the pattern provided by @chrisdunelm here. I am okay with adding `isSupported` into the feature config but I prefer to make the change in a different PR.

File: src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java [hunk 5]
```diff
@@ -238,6 +240,12 @@ public StaticLangApiMethodView generateCallableMethod(MethodTransformerContext c
         !ServiceMessages.s_isEmptyType(context.getMethod().getOutputType()));
     methodViewBuilder.isPageStreaming(false);
 
+    MethodConfig methodConfig = context.getMethodConfig();
+    if (methodConfig.isGrpcStreaming()) {
+      GrpcStreamingConfig grpcStreaming = methodConfig.getGrpcStreaming();
+      methodViewBuilder.streamingType(grpcStreaming.getType());
+    }
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java]: Maybe it would be easier to have this if condition elsewhere, so that we could just always set streamingType (and set it to None if there is no streaming)

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java [hunk 6]
```diff
@@ -249,7 +243,8 @@ private void addXSettingsImports(SurfaceTransformerContext context) {
     ModelTypeTable typeTable = context.getTypeTable();
     typeTable.saveNicknameFor("com.google.api.gax.core.ConnectionSettings");
     typeTable.saveNicknameFor("com.google.api.gax.core.RetrySettings");
-    typeTable.saveNicknameFor("com.google.api.gax.grpc.ApiCallSettings");
+    typeTable.saveNicknameFor("com.google.api.gax.grpc.UnaryApiCallSettings");
+    typeTable.saveNicknameFor("com.google.api.gax.grpc.StreamingCallSettings");
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: The StreamingCallSettings won't always be used. The existing pattern is to add the import only in the branches that generate code depending on the type.

File: src/main/java/com/google/api/codegen/viewmodel/StaticLangXSettingsView.java [hunk 7]
```diff
@@ -37,6 +39,16 @@
...
 
+  public List<ApiCallSettingsView> unaryCallSettings() {
+    ArrayList<ApiCallSettingsView> unaryCallSettings = new ArrayList<>();
+    for (ApiCallSettingsView settingsView : callSettings()) {
+      if (!settingsView.isStreaming()) {
+        unaryCallSettings.add(settingsView);
+      }
+    }
+    return unaryCallSettings;
+  }
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/viewmodel/StaticLangXSettingsView.java]: Don't put logic like this in view models

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline [hunk 8]
```diff
@@ -1643,12 +1655,124 @@ public class LibraryServiceApi implements AutoCloseable {
...
    */
-  public final ApiCallable<UpdateBookIndexRequest, Empty> updateBookIndexCallable() {
+  public final UnaryApiCallable<UpdateBookIndexRequest, Empty> updateBookIndexCallable() {
     return updateBookIndexCallable;
   }
 
   // AUTO-GENERATED DOCUMENTATION AND METHOD
   /**
+   * Test server streaming
+   * gRPC streaming methods don't have HTTP equivalent and need not have google.api.http option.
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: -> "gRPC streaming methods don't have an HTTP equivalent and don't need to have the google.api.http option."

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline [hunk 9]
```diff
@@ -263,6 +273,27 @@ public class LibraryServiceSettings extends ServiceApiSettings {
   }
 
   /**
+   * Returns the object with the settings used for calls to streamShelves.
+   */
+  public StreamingCallSettings<StreamShelvesRequest, StreamShelvesResponse> streamShelvesSettings() {
+    return streamShelvesSettings;
+   }
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Why is the indentation of this bracket wrong?

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline [hunk 10]
```diff
@@ -572,6 +606,9 @@ public class LibraryServiceSettings extends ServiceApiSettings {
     private final SimpleCallSettings.Builder<AddCommentsRequest, Empty> addCommentsSettings;
     private final SimpleCallSettings.Builder<GetBookFromArchiveRequest, Book> getBookFromArchiveSettings;
     private final SimpleCallSettings.Builder<UpdateBookIndexRequest, Empty> updateBookIndexSettings;
+    private StreamingCallSettings.Builder<StreamShelvesRequest, StreamShelvesResponse> streamShelvesSettings;
+    private StreamingCallSettings.Builder<StreamBooksRequest, Book> streamBooksSettings;
+    private StreamingCallSettings.Builder<DiscussBookRequest, Comment> discussBookSettings;
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: make final

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline [hunk 11]
```diff
@@ -658,7 +701,7 @@ public class LibraryServiceSettings extends ServiceApiSettings {
 
       addLabelSettings = SimpleCallSettings.newBuilder(LabelerGrpc.METHOD_ADD_LABEL);
 
-      methodSettingsBuilders = ImmutableList.<ApiCallSettings.Builder>of(
+      unaryMethodSettingsBuilders = ImmutableList.<ApiCallSettings.Builder>of(
```
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Why no streamingMethodSettingsBuilders? I see that as a consequence, you don't support applyToAllApiMethods for streaming methods.

File: src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java (hunk 2 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java]: @jmuk This follows Jun's grpc streaming design and implementation of Python. Both page-streaming and grpc-streaming shares the same the config proto for now. But I would prefer to have a separate proto for grpc-streaming. With that said, if we agree that we should use a separate proto, I can add a TODO here and make the change in a separate PR.

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline (hunk 11 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Right, applyToAllApiMethods does not support streaming methods since right now we do not have any settings for streaming methods yet.

File: src/main/java/com/google/api/codegen/config/MethodConfig.java (hunk 3 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/config/MethodConfig.java]: It was only used within this class but I found there are other places can reuse this method. Now I make it public. Done.

File: src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java (hunk 4 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java]: This is actually not used...removed.

File: src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java (hunk 5 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java]: I am not sure if I understand this comment. I moved this if block into the `MethodConfig` class.

File: src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java (hunk 6 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/java/JavaGapicSurfaceTransformer.java]: Right, I was thinking we now have the import cleaner so I just did not implement the import logic. The issue is that the branch that generates the code is not language specific. (e.g. StreamingCallSettings are generated in ApiCallableTransformer). 

My fix is to add an extra helper method to determine we need to add such imports.

File: src/main/java/com/google/api/codegen/viewmodel/StaticLangXSettingsView.java (hunk 7 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/viewmodel/StaticLangXSettingsView.java]: Done.

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline (hunk 8 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Done.

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline (hunk 9 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: That's the typo in the snip file. Fixed.

File: src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline (hunk 10 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_settings_library.baseline]: Done

**Comment by shinfan** [comment]
PTAL

File: src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java [hunk 12]
```diff
@@ -79,7 +80,10 @@ public ApiCallableTransformer() {
     apiCallableBuilder.memberName(context.getNamer().getSettingsMemberName(method));
     apiCallableBuilder.settingsFunctionName(context.getNamer().getSettingsFunctionName(method));
 
-    if (methodConfig.isBundling()) {
+    if (methodConfig.isGrpcStreaming()) {
+      apiCallableBuilder.type(ApiCallableType.StreamingApiCallable);
+      apiCallableBuilder.streamingType(methodConfig.getGrpcStreaming().getType());
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java]: If this is just gRPC streaming, I'd prefer the view's attribute name to be called `grpcStreamingType` (same for the other places where `streamingType` occurs).

File: src/main/java/com/google/api/codegen/transformer/FeatureConfig.java [hunk 13]
```diff
@@ -36,4 +36,9 @@ public boolean useResourceNameFormatOption(Field field) {
   public boolean enableMixins() {
     return false;
   }
+
+  /** Returns true if streaming APIs are supported. */
+  public boolean enableStreaming() {
```
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/FeatureConfig.java]: I would prefer "enableGrpcStreaming".

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline (hunk 8 above)
Comment (garrettjonesgoogle) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Doesn't look done.

File: src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java (hunk 2 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java]: Yeah, add a TODO. I want either of the following:

* Rename the config proto to StreamingConfigProto
* Create a separate GrpcStreamingConfigProto

It's really confusing to have one name specifically for one case but used for both.

File: src/main/java/com/google/api/codegen/config/MethodConfig.java (hunk 3 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/config/MethodConfig.java]: Ahh, I misread this - you had the `static` before the `private`; it should have been the other order.

File: src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java (hunk 5 above)
Comment (garrettjonesgoogle) [inline src/main/java/com/google/api/codegen/transformer/ApiMethodTransformer.java]: That's what I was going for :-)

File: src/main/java/com/google/api/codegen/transformer/FeatureConfig.java (hunk 13 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/FeatureConfig.java]: Done

File: src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java (hunk 12 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/transformer/ApiCallableTransformer.java]: Done

File: src/test/java/com/google/api/codegen/testdata/java_main_library.baseline (hunk 8 above)
Comment (shinfan) [inline src/test/java/com/google/api/codegen/testdata/java_main_library.baseline]: Oh sorry somehow I thought it was fixed. Fixed now.

File: src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java (hunk 2 above)
Comment (shinfan) [inline src/main/java/com/google/api/codegen/config/GrpcStreamingConfig.java]: Picked `Create a separate GrpcStreamingConfigProto`. TODO added.

**Comment by garrettjonesgoogle** [comment]
LGTM (after tests pass)
